mllib/src/test/scala/org/apache/spark/ml/tree/impl/RandomForestSuite.scala

```diff
@@ -631,6 +651,160 @@ class RandomForestSuite extends SparkFunSuite with MLlibTestSparkContext {
+             root
+      left1       right1
+                left2  right2
+
+      pred(left1)  = 0
+      pred(left2)  = 1
+      pred(right2) = 0
+     */
+    assert(dt.rootNode.numDescendants === 4)
+    assert(dt.rootNode.subtreeDepth === 2)
+
+    assert(dt.rootNode.isInstanceOf[InternalNode])
+
+    // left 1 prediction test
+    assert(dt.rootNode.asInstanceOf[InternalNode].leftChild.prediction === 0)
+
+    val right1 = dt.rootNode.asInstanceOf[InternalNode].rightChild
+    assert(right1.isInstanceOf[InternalNode])
+
+    // left 2 prediction test
+    assert(right1.asInstanceOf[InternalNode].leftChild.prediction === 1)
+    // right 2 prediction test
+    assert(right1.asInstanceOf[InternalNode].rightChild.prediction === 0)
+  }
+
+  test("[SPARK-3159] tree model redundancy - multiclass classification") {
+    val numClasses = 4
+
+    val strategy = new OldStrategy(algo = OldAlgo.Classification, impurity = Gini, maxDepth = 4,
+      numClasses = numClasses, maxBins = 32)
+
+    val dt = buildRedundantDecisionTree(numClasses, 20, strategy = strategy)
+
+    /* Expected tree structure tested below:
+                    root
+            left1         right1
+        left2  right2  left3  right3
+
+      pred(left2)  = 0
+      pred(right2)  = 1
+      pred(left3) = 2
+      pred(right3) = 1
+     */
+    assert(dt.rootNode.numDescendants === 6)
+    assert(dt.rootNode.subtreeDepth === 2)
+
+    assert(dt.rootNode.isInstanceOf[InternalNode])
+
+    val left1 = dt.rootNode.asInstanceOf[InternalNode].leftChild
+    val right1 = dt.rootNode.asInstanceOf[InternalNode].rightChild
+
+    assert(left1.isInstanceOf[InternalNode])
+
+    // left 2 prediction test
+    assert(left1.asInstanceOf[InternalNode].leftChild.prediction === 0)
+    // right 2 prediction test
+    assert(left1.asInstanceOf[InternalNode].rightChild.prediction === 1)
+
+    assert(right1.isInstanceOf[InternalNode])
+
+    // left 3 prediction test
+    assert(right1.asInstanceOf[InternalNode].leftChild.prediction === 2)
+    // right 3 prediction test
+    assert(right1.asInstanceOf[InternalNode].rightChild.prediction === 1)
+  }
+
+  test("[SPARK-3159] tree model redundancy - regression") {
+    val numClasses = 2
+
+    val strategy = new OldStrategy(algo = OldAlgo.Regression, impurity = Variance,
+      maxDepth = 3, maxBins = 10, numClasses = numClasses)
+
+    val dt = buildRedundantDecisionTree(numClasses, 20, strategy = strategy)
+
+    /* Expected tree structure tested below:
+                                root
+                    1                             2
+           1_1             1_2             2_1         2_2
+      1_1_1   1_1_2   1_2_1   1_2_2   2_1_1   2_1_2
+
+      pred(1_1_1)  = 0.5
+      pred(1_1_2)  = 0.0
+      pred(1_2_1)  = 0.0
+      pred(1_2_2)  = 0.25
+      pred(2_1_1)  = 1.0
+      pred(2_1_2)  = 0.6666666666666666
+      pred(2_2)    = 0.5
+     */
+
+    assert(dt.rootNode.numDescendants === 12)
```

Comment:
Ok, trying to understand these tests. From what I can tell, you've written a data generator that generates random points, and, somewhat by chance, generates redundant tree nodes if the tree is not pruned. Your relying on the random seed to give you a tree which should have exactly 12 descendants after pruning. 

I think these may be overly complicated. IMO, we just need to test the situation that causes this by creating simple dataset that can improve the impurity by splitting, but which does not change the prediction. For example, for the Gini impurity you might have the following:

```scala
    val data = Array(
      LabeledPoint(0.0, Vectors.dense(1.0)),
      LabeledPoint(0.0, Vectors.dense(1.0)),
      LabeledPoint(0.0, Vectors.dense(1.0)),
      LabeledPoint(0.0, Vectors.dense(0.0)),
      LabeledPoint(0.0, Vectors.dense(0.0)),
      LabeledPoint(1.0, Vectors.dense(0.0))
    )
    val rdd = sc.parallelize(data)
    val strategy = new OldStrategy(algo = OldAlgo.Regression, impurity = Variance, maxDepth = 4,
      numClasses = 0, maxBins = 32)
    val tree = RandomForest.run(rdd, strategy, numTrees = 1, featureSubsetStrategy = "auto",
      seed = 42, instr = None).head
    assert(tree.rootNode.numDescendants === 0)
```
Before this patch, you'd get a tree with two descendants. Actually, I think it might be nice to introduce a `prune` parameter just for testing, then you can do the following:

```scala
    val data = Array(
      LabeledPoint(0.0, Vectors.dense(1.0)),
      LabeledPoint(0.0, Vectors.dense(1.0)),
      LabeledPoint(0.0, Vectors.dense(1.0)),
      LabeledPoint(0.0, Vectors.dense(0.0)),
      LabeledPoint(0.0, Vectors.dense(0.0)),
      LabeledPoint(1.0, Vectors.dense(0.0))
    )
    val rdd = sc.parallelize(data)
    val strategy = new OldStrategy(algo = OldAlgo.Regression, impurity = Variance, maxDepth = 4,
      numClasses = 0, maxBins = 32)
    val prunedTree = RandomForest.run(rdd, strategy, numTrees = 1, featureSubsetStrategy = "auto",
      seed = 42, prune = true, instr = None).head
    val unprunedTree = RandomForest.run(rdd, strategy, numTrees = 1, featureSubsetStrategy = "auto",
      seed = 42, prune = false, instr = None).head
    assert(prunedTree.rootNode.numDescendants === 0)
    assert(unprunedTree.rootNode.numDescendants === 2)
```
That's a sanity check that the patch has actually made the difference. WDYT?

Reply:
The tree tests are already so long and complicated that I think it's important to simplify where possible. These tests are useful as they are, but it probably won't be obvious why/how they work to future devs. Also, if we can avoid adding data generation code, that would be nice (there's already tons of code like that laying around the test suites).

Reply:
I agree on the whole line, and here are the modifications I have made to align to the comments: 
- I have simplified the tests as much as possible (minimal input data, resulting into minimal trees)
- I have removed the data generator methods (now the input data are explicitly given)
- Conditions to be tested are simple (just the number of nodes as you suggested), and they test the pruned vs unpruned version
- I have added the extra debugging parameter _prune_ (to _RandomForest.run(...)_ and _Node.toNode()_)
- Now that we have _prune_, I have reverted back the adaptation of "Use soft prediction for binary classification with ordered categorical features", because now it is sufficient to add `prune = false` to the invocation of _RandomForest.run(...)_)